core/domain/email_services.py

```diff
@@ -69,3 +78,219 @@ def get_feedback_thread_reply_info_by_user_and_thread_ids(user_id, thread_id):
     if model is None:
         return None
     return get_feedback_thread_reply_info_from_model(model)
+
+
+def get_incoming_email_address(reply_to_id):
+    """Gets the incoming email address. The client is responsible for recording
+    any audit logs.
+
+    Args:
+        reply_to_id: str. The unique id of the sender.
+
+    Returns:
+        str. The email address of the sender.
+    """
+    return 'reply+%s@%s' % (reply_to_id, feconf.INCOMING_EMAILS_DOMAIN_NAME)
+
+
+def is_email_valid(email_address):
+    """Determines whether an email address is invalid.
+
+    Args:
+        email_address: str. Email address to check.
+
+    Returns:
+        bool. Whether specified email address is valid.
+    """
+    if not isinstance(email_address, python_utils.BASESTRING):
+        return False
+
+    stripped_address = email_address.strip()
+    if not stripped_address:
+        return False
+    regex = r'^.+@(\[?)[a-zA-Z0-9-.]+.([a-zA-Z]{2,3}|[0-9]{1,3})(]?)$'
+    return re.search(regex, email_address)
+
+
+def is_sender_email_valid(sender_email):
+    """Gets the sender_email address and validates it is of the form
+    'SENDER_NAME <SENDER_EMAIL_ADDRESS>' or 'email_address'.
+
+    Args:
+        sender_email: str. The email address of the sender.
+
+    Returns:
+        bool. Whether the sender_email is valid.
+    """
+    sender_email_with_only_first_name_regex = (
+        r'^[a-zA-Z._]+ <.+@(\[?)[a-zA-Z0-9-.]+.([a-zA-Z]{2,3}|[0-9]{1,3})' +
+        '(]?)>$')
+    sender_email_with_full_name_regex = (
+        r'^[a-zA-Z._]+ [a-zA-Z._]+ <.+@(\[?)[a-zA-Z0-9-.]+.([a-zA-Z]{2,3}' +
+        r'|[0-9]{1,3})(]?)>$')
+    return is_email_valid(sender_email) or (
+        re.search(sender_email_with_only_first_name_regex, sender_email)) or (
+            re.search(sender_email_with_full_name_regex, sender_email))
+
+
+def send_mail(
+        sender_email, recipient_email, subject, plaintext_body,
+        html_body, bcc_admin=False, reply_to_id=None):
+    """Sends an email using mailgun api.
+
+    In general this function should only be called from
+    email_manager._send_email().
+
+    Args:
+        sender_email: str. the email address of the sender. This should be in
+            the form 'SENDER_NAME <SENDER_EMAIL_ADDRESS>'.
+        recipient_email: str. the email address of the recipient.
+        subject: str. The subject line of the email.
+        plaintext_body: str. The plaintext body of the email.
+        html_body: str. The HTML body of the email. Must fit in a datastore
+            entity.
+        bcc_admin: bool. Whether to bcc feconf.ADMIN_EMAIL_ADDRESS on the email.
+        reply_to_id: str. The unique id of the sender.
+
+    Raises:
+        Exception: if the configuration in feconf.py forbids emails from being
```

Comment:
Exception: The configuration ...

(Phrase others similarly. Also there are other exceptions which do not seem to be recorded here.)